grr/server/grr_response_server/gui/ui/lib/api_translation/client.ts

```diff
@@ -30,6 +32,124 @@ function createClientLabel(label: ApiClientLabel): ClientLabel {
+    node: apiUname.node,
+    release: apiUname.release,
+    version: apiUname.version,
+    machine: apiUname.machine,
+    kernel: apiUname.kernel,
+    fqdn: apiUname.fqdn,
+    installDate: createOptionalDate(apiUname.installDate),
+    libcVer: apiUname.libcVer,
+    architecture: apiUname.architecture,
+  }
+}
+
+function createUser(apiUser: ApiUser): User {
+  return {
+    username: apiUser.username,
+    fullName: apiUser.fullName,
+    lastLogon: createOptionalDate(apiUser.lastLogon),
+    uid: apiUser.uid,
+    gid: apiUser.gid,
+    shell: apiUser.shell,
+    homedir: apiUser.homedir,
+  }
+}
+
+function createNetworkAddress(apiNetAddress: ApiNetworkAddress): NetworkAddress {
+  if (!apiNetAddress.addressType) throw new Error('addressType attribute is missing.');
+  if (!apiNetAddress.packedBytes) throw new Error('packedBytes attribute is missing.');
+
+  let addressType = 'IPv4'
+  if (apiNetAddress.addressType === 'INET6') {
+    addressType = 'IPv6';
+  }
+
+  const addressBytes = decodeBase64(apiNetAddress.packedBytes);
+
+  return {
+    addressType: addressType,
+    ipAddress: (addressType === 'IPv4') ? createIpv4Address(addressBytes) : createIpv6Address(addressBytes),
+  }
+}
+
+function createNetworkInterface(apiInterface: ApiInterface): NetworkInterface {
+  if (!apiInterface.macAddress) throw new Error('macAddress attribute is missing.');
+  if (!apiInterface.ifname) throw new Error('ifname attribute is missing.');
+
+  return {
+    macAddress: createMacAddress(decodeBase64(apiInterface.macAddress)),
+    interfaceName: apiInterface.ifname,
+    addresses: (apiInterface.addresses || []).map(createNetworkAddress),
+  }
+}
+
+function createOptionalWindowsVolume(volume?: ApiWindowsVolume): WindowsVolume | undefined {
+  if (volume === undefined) {
+    return undefined;
+  }
+
+  return {
+    attributes: volume.attributesList,
+    driveLetter: volume.driveLetter,
+    driveType: volume.driveType,
+  }
+}
+
+function createOptionalUnixVolume(volume?: ApiUnixVolume): UnixVolume | undefined {
+  if (volume === undefined) {
+    return undefined;
+  }
+
+  return {
+    mountPoint: volume.mountPoint,
+    mountOptions: volume.options,
+  }
+}
+
+function createStorageVolume(apiVolume: ApiVolume): StorageVolume {
+  let totalSize = 0;
+  let freeSpace = 0;
+  if (apiVolume.bytesPerSector && apiVolume.sectorsPerAllocationUnit) {
+    if (apiVolume.totalAllocationUnits) {
+      totalSize = apiVolume.bytesPerSector *
+          apiVolume.sectorsPerAllocationUnit * apiVolume.totalAllocationUnits;
+    }
+
+    if (apiVolume.actualAvailableAllocationUnits) {
+      freeSpace = apiVolume.bytesPerSector *
+          apiVolume.sectorsPerAllocationUnit *
+          apiVolume.actualAvailableAllocationUnits;
+    }
+  }
+
+  return {
+    name: apiVolume.name,
+    devicePath: apiVolume.devicePath,
+    fileSystemType: apiVolume.fileSystemType,
+    bytesPerSector: apiVolume.bytesPerSector,
+    totalSize: totalSize,
```

Comment:
`totalSize` and `freeSpace` are `0` if `bytesPerSector` or `sectorsPerAllocationUnit` are undefined. Is this intended?

Reply:
Yes, would be better to initialize them with `undefined` and not show them in the UI if they are not available?

Reply:
Yes.